File: backend/app/schemas/knowledge.py
"""知识库 Pydantic schemas"""

from pydantic import BaseModel
from typing import Optional


class WorldSettingResponse(BaseModel):
    id: int
    project_id: int
    core_concept: Optional[str] = None
    tiered_settings: dict = {}
    key_locations: list = []

    class Config:
        from_attributes = True


class WorldSettingUpdate(BaseModel):
    core_concept: Optional[str] = None
    tiered_settings: Optional[dict] = None
    key_locations: Optional[list] = None


class StyleConstraintsResponse(BaseModel):
    id: int
    project_id: int
    taboo_words: list = []
    forbidden_patterns: list = []
    style_anchor: Optional[str] = None
    abstract_rules: list = []

    class Config:
        from_attributes = True


class StyleConstraintsUpdate(BaseModel):
    taboo_words: Optional[list] = None
    forbidden_patterns: Optional[list] = None
    style_anchor: Optional[str] = None
    abstract_rules: Optional[list] = None


class PlotBlockResponse(BaseModel):
    id: int
    project_id: int
    title: str
    questions_to_answer: list = []
    questions_to_raise: list = []
    must_happen: list = []
    expected_mood: Optional[str] = None
    chapter_start: Optional[int] = None
    chapter_end: Optional[int] = None
    completion_summary: Optional[str] = None

    class Config:
        from_attributes = True


class ForeshadowingResponse(BaseModel):
    id: int
    project_id: int
    content: str
    level: str = "hint"
    appearance_count: int = 1
    status: str = "active"
    planted_chapter: Optional[int] = None
    expected_resolve_chapter: Optional[int] = None
    resolved_chapter: Optional[int] = None
    related_characters: list = []

    class Config:
        from_attributes = True


class TimelineEntryResponse(BaseModel):
    id: int
    project_id: int
    chapter_number: int
    summary: Optional[str] = None
    causal_chain: Optional[str] = None
    rhythm_score: int = 3
    tension_score: int = 3
    emotion_score: int = 3
    emotion_tag: Optional[str] = None

    class Config:
        from_attributes = True


class StyleSnapshotResponse(BaseModel):
    id: int
    project_id: int
    chapter_number: int
    paragraph_count: int = 0
    avg_paragraph_length: float = 0.0
    dialogue_ratio: float = 0.0
    avg_sentence_length: float = 0.0

    class Config:
        from_attributes = True

File: frontend/src/types/knowledge.ts
// 知识库相关类型定义

export interface PlotBlock {
  id: number
  project_id: number
  title: string
  questions_to_answer: string[]
  questions_to_raise: string[]
  must_happen: string[]
  expected_mood: string | null
  chapter_start: number | null
  chapter_end: number | null
  completion_summary: string | null
}

export interface PlotBlockUpdate {
  title?: string
  questions_to_answer?: string[]
  questions_to_raise?: string[]
  must_happen?: string[]
  expected_mood?: string
  chapter_start?: number
  chapter_end?: number
  completion_summary?: string
}

export interface Subplot {
  id: number
  project_id: number
  name: string
  characters: string[]
  current_status: string
  raised_in_chapter: number | null
  planned_intersection_chapter: number | null
  expected_resolution_chapter: number | null
}

export interface SubplotCreate {
  name: string
  characters?: string[]
  current_status?: string
  raised_in_chapter?: number | null
  planned_intersection_chapter?: number | null
  expected_resolution_chapter?: number | null
}

export interface SubplotUpdate {
  name?: string
  characters?: string[]
  current_status?: string
  raised_in_chapter?: number | null
  planned_intersection_chapter?: number | null
  expected_resolution_chapter?: number | null
}

export interface Foreshadowing {
  id: number
  project_id: number
  content: string
  level: string
  appearance_count: number
  status: string
  planted_chapter: number | null
  expected_resolve_chapter: number | null
  resolved_chapter: number | null
  related_characters: string[]
}

export interface ForeshadowingUpdate {
  content?: string
  level?: string
  status?: string
  planted_chapter?: number | null
  expected_resolve_chapter?: number | null
  resolved_chapter?: number | null
  related_characters?: string[]
}

export interface TimelineEntry {
  id: number
  project_id: number
  chapter_number: number
  summary: string | null
  causal_chain: string | null
  rhythm_score: number
  tension_score: number
  emotion_score: number
  emotion_tag: string | null
}

export interface StyleSnapshot {
  id: number
  project_id: number
  chapter_number: number
  paragraph_count: number
  avg_paragraph_length: number
  dialogue_ratio: number
  avg_sentence_length: number
  ai_marker_density?: number
  sentence_variety?: number
}

export interface WorldSetting {
  id: number
  project_id: number
  core_concept: string | null
  tiered_settings: Record<string, string[]>
  key_locations: string[]
}

export interface StyleConstraints {
  id: number
  project_id: number
  taboo_words: string[]
  forbidden_patterns: string[]
  style_anchor: string | null
  abstract_rules: string[]
}
